lib/prom_ex/plugins/plug_cowboy.ex

```diff
@@ -0,0 +1,185 @@
+            measurement: :resp_duration,
+            description: "The time it takes for the application to send the HTTP response.",
+            reporter_options: [
+              buckets: exponential!(1, 2, 12)
+            ],
+            drop: drop_ignored(ignore_routes),
+            tag_values: &get_tags/1,
+            tags: http_metrics_tags,
+            unit: {:native, :millisecond}
+          ),
+          distribution(
+            metric_prefix ++ [:http, :request_body, :duration, :milliseconds],
+            event_name: cowboy_stop_event,
+            measurement: :req_body_duration,
+            description: "The time it takes for the application to receive the HTTP request body.",
+            reporter_options: [
+              buckets: exponential!(1, 2, 12)
+            ],
+            tag_values: &get_tags/1,
+            tags: http_metrics_tags,
+            unit: {:native, :millisecond}
+          ),
+
+          # Capture response payload size information
+          distribution(
+            metric_prefix ++ [:http, :response, :size, :bytes],
+            event_name: cowboy_stop_event,
+            measurement: :resp_body_length,
+            description: "The size of the HTTP response payload.",
+            reporter_options: [
+              buckets: exponential!(1, 4, 12)
+            ],
+            drop: drop_ignored(ignore_routes),
+            tag_values: &get_tags/1,
+            tags: http_metrics_tags,
+            unit: :byte
+          ),
+
+          # Capture the number of requests that have been serviced
+          counter(
+            metric_prefix ++ [:http, :requests, :total],
+            event_name: cowboy_stop_event,
+            description: "The number of requests have been serviced.",
+            drop: drop_ignored(ignore_routes),
+            tag_values: &get_tags/1,
+            tags: http_metrics_tags
+          )
+        ]
+      )
+    end
+
+    defp get_tags(%{resp_status: resp_status, req: %{method: method, path: path}}) do
+      case get_http_status(resp_status) do
+        status when is_binary(status) ->
+          %{
+            status: status,
+            method: method,
+            path: path
+          }
+
+        :undefined ->
+          %{
+            status: :undefined,
+            method: method,
+            path: path
+          }
+
+        nil ->
+          Logger.warn("Cowboy failed to provide valid response status #{inspect(resp_status)}")
+          %{}
+      end
+    end
+
+    defp get_http_status(resp_status) when is_integer(resp_status) do
+      to_string(resp_status)
+    end
+
+    defp get_http_status(resp_status) when is_bitstring(resp_status) do
+      [code | _rest] = String.split(resp_status)
+      code
+    end
+
+    defp get_http_status(nil) do
+      nil
+    end
+
+    defp get_http_status(_resp_status) do
+      :undefined
+    end
+
+    defp drop_ignored(ignored_routes) do
+      fn
+        %{req: %{path: path}} ->
```

Comment:
Create the MapSet outside of the returned function that way the MapSet is not generated from scratch on every request.